### ISSUE TITLE
Create tox tests and add CI

### ISSUE THREAD

**Issue by np-8** [np-8/wakepy#30]
Create basic tox tests and add CI (github actions?).

**Comment by np-8** [comment]
Started working on this. Some notes:

### (1) No D-Bus -> All tests will fail
In GitHub actions, the runner used is very bare bones Ubuntu 20.04 / 22.04. There is (unsurprisingly) no Desktop Environment, and no session D-Bus. Therefore, the used D-Bus methods will fail. From a recent [job output](https://github.com/np-8/wakepy/actions/runs/5181748273/jobs/9337678415)

Checking for dbus process (only system dbus started):
```
Run ps -ef | grep dbus | grep -v grep
message+     656       1  0 20:48 ?        00:00:00 /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
```

and tests failing because inhibition would use session dbus:

```
>               ) from e
E               NotImplementedError: DBUS_SESSION_BUS_ADDRESS environment variable not set! If running in subprocess, make sure to pass the DBUS_SESSION_BUS_ADDRESS environment variable.

.tox/linux-py37/lib/python3.7/site-packages/wakepy/_deprecated/_linux.py:45: NotImplementedError
```

I'm not sure if it is possible to install a Desktop Environment, launch the session D-Bus, and get the test working. Perhaps. But that only works here. **All the other packages using wakepy are probably also failing all their CI tests because of this.**.  So clearly, this needs some good solution for CI of _other packages_ using wakepy.  Something like:

```python

with long_task_mode() as m:
    if not m.success:
        ... 
    else:
        # do something that might take a while.
```

would allow to gracefully exit if a inhibitor lock could not be taken. Another possibility would be to provide some mock context manager with would "always succeed" (for patching in tests). Not sure if that would help.

### (2) Python 3.6 experience

Tried to run the tests with Python 3.6 and they failed (even faster) since there is no support for 

```python
from __future__ import annotations
```

That's used quite a lot to make development nicer with type annotations. Not sure if it's worth it to look at python 3.6 some day, but at least for now it seems that I could try to make wakepy support 3.7 and above (was previously >=3.8).

**Comment by np-8** [comment]
### wakepy in CI?

#### User's perpective
Thinking about this problem again. If someone is writing a library `superlib` or application `my-awesome-app`, and using wakepy to keep it alive with code similar to:

```python
with long_task_mode() as m:
    if not m.success:
        bail()
    else:
        do_the_thingy()
``` 

they would, in most CI pipelines, not get `do_the_thingy` called, but  `bail`. That would mean that the application/library might not get tested properly. It could also mean that the code simply raises an exception. This is in the case if wakepy does not happen to support inhibiting the suspend/idle mode/whatever on whichever linux server that code is being ran. In most cases, there will be no Desktop Environment, so no GNOME/KDE/Xfce to respond to any D-Bus calls. While adding support to any possible system is in at least in theory in scope, there should be better solution for this.

#### Possible solution: env variable `WAKEPY_FAKE_SUCCESS`

What I now can come with from the top of my head would be:
- Wakepy first tries all methods (or all given methods) to enter a specified mode
- If that fails, look if there is environment variable `WAKEPY_FAKE_SUCCESS` set to "yes". 
    - if "yes", make the return value of the context manager `__enter__` to have `.success` of  `True`, even if it would have failed. There could be also `._real_success` or something which would be `False`.
    - If "no", continue normally; `m.success` is `False`.

Setting environment variables for CIs is quite easy, and it does not require any application/library code changes, nor it requires any changes / mocking in tests. There could even be possibility of `WAKEPY_INSTANT_FAKE_SUCCESS`  which would short-circuit to success (and not try any methods).

**Comment by np-8** [comment]
#### Wakepy development perspective

From the point of view of wakepy development, it would be nice to have tests on runners with some specific systems and Desktop Environments +  programs installed. Let's see if that's possible at all on GitHub actions. 

In any case, it would be good to systematically take notes on behavior of different methods on different systems and store them. The testing in any CI solution will be quite limited, and will be focused on preventing obvious bugs in the python code and alarming if something is not available in Python 3.7 or the used OS.

**Comment by np-8** [comment]
There's now some basic tests with tox + GitHub actions. These will test all the python parts with linux, windows and macos --> Closed this. 

It's a bit different thing to start testing that the OS / Desktop Evironment / other commands called, like the D-Bus `Inhibit` method of `org.freedesktop.ScreenSaver` actually does what it is supposed to do. I think that's something what does not need _automatic_ tests in wakepy, but manual tests and documentation (how different methods work on different Desktop Environments).
